api/pkg/s3/object-utils.go

```diff
@@ -0,0 +1,100 @@
+/*
+ * Minio Cloud Storage, (C) 2015, 2016 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package s3
+
+import (
+	"regexp"
+	"strconv"
+	"strings"
+	"unicode/utf8"
+	"fmt"
+)
+
+// validBucket regexp.
+var validBucket = regexp.MustCompile(`^[a-z0-9][a-z0-9\.\-]{1,61}[a-z0-9]$`)
+
+// IsValidBucketName verifies a bucket name in accordance with Amazon's
+// requirements. It must be 3-63 characters long, can contain dashes
+// and periods, but must begin and end with a lowercase letter or a number.
+// See: http://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html
+func isValidBucketName(bucketName string) bool {
+	if !validBucket.MatchString(bucketName) {
+		return false
+	}
+	// make sure there're no continuous dots
+	if strings.Contains(bucketName, "..") {
+		return false
+	}
+	// make sure it's not an IP address
+	split := strings.Split(bucketName, ".")
+	if len(split) == 4 {
+		for _, p := range split {
+			n, err := strconv.Atoi(p)
+			if err == nil && n >= 0 && n <= 255 {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// IsValidObjectName verifies an object name in accordance with Amazon's
+// requirements. It cannot exceed 1024 characters and must be a valid UTF8
+// string.
+// Some characters require special handling:
+// & $ @ = ; : + (space) , ?
+// and ASCII ranges 0x00-0x1F(0-31 decimal) and 7F(127 decimal)
+// Some characters to avoid:
+// \ { ^ } % ` [ ] ' " < > ~ # |
+// and non-printable ASCII characters(128-255 decimal)
+//
+// As in YIG, we PROHIBIT ALL the characters listed above
+// See http://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html
+func isValidObjectName(objectName string) bool {
```

Comment:
same here, why not call an AWS API ? each cloud provider will have different valid rules, right ?

Reply:
@rhsakarpos the function isValidObjectName here is migrated from yig. and because api layer provided by s3 in multi-cloud must be compatible with s3 protocol, and this function is implemented according to s3 protocol, so, this function works fine here.

Reply:
sorry, i do not agree with this. imho, we should NOT have any cloud vendor specific code in OSDS

Reply:
@rhsakarpos these two kinds of functions are ported from yig and they are not vendor specific code.

Reply:
@hopkings2008 this is S3 logic to check the validity of bucket name or object name. please refer to previous comment response, thanks

Reply:
@hopkings2008  imho, we do NOT need any check (with code) in OSDS. We simply call the cloud provider API...S3 or Azure or whatever cloud backend.

Reply:
hi @rhsakarpos this is similar to [this](https://github.com/opensds/multi-cloud/pull/191#discussion_r338526608)